crates/ark/src/lsp/completions/sources/composite.rs

```diff
@@ -78,16 +101,81 @@ pub(crate) fn get_completions(
         )?;
     }
 
-    // Remove duplicates
-    let mut uniques = HashSet::new();
-    completions.retain(|x| uniques.insert(x.label.clone()));
+    // Simplify to plain old CompletionItems and sort them
+    let completions = finalize_completions(&completions);
+
+    Ok(Some(completions))
+}
+
+fn push_completions<S>(
+    source: S,
+    completion_context: &CompletionContext,
+    completion_map: &mut HashMap<CompletionKey, CompletionWithSource>,
+) -> anyhow::Result<()>
+where
+    S: CompletionSource,
+{
+    let source_name = source.name();
+
+    if let Some(source_completions) = collect_completions(source, completion_context)? {
+        for item in source_completions {
+            let key = CompletionKey::from(&item);
+
+            if let Some(sourced_item) = completion_map.get_mut(&key) {
+                // Item already exists, just add this source
+                sourced_item.sources.push(source_name.to_string());
+                log::debug!(
+                    "Completion '{}' contributed by multiple sources: {:?}",
+                    key.label,
+                    sourced_item.sources
+                );
+            } else {
+                // New item
+                let mut sourced_item = CompletionWithSource::default();
+                sourced_item.item = item;
+                sourced_item.sources.push(source_name.to_string());
+                completion_map.insert(key, sourced_item);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Produce and sort plain CompletionItems, with source information
+/// stashed in `data` if needed
+fn finalize_completions(
+    completion_map: &HashMap<CompletionKey, CompletionWithSource>,
+) -> Vec<CompletionItem> {
+    let mut items: Vec<CompletionItem> = completion_map
+        .values()
+        .map(|sourced_item| {
+            let mut item = sourced_item.item.clone();
+
+            // Store source information in the `data` field as a simple string
+            if !sourced_item.sources.is_empty() {
+                let sources_string = sourced_item.sources.join(",");
+                item.data = Some(serde_json::Value::String(sources_string.clone()));
+                log::debug!(
```

Comment:
```suggestion
                log::trace!(
```

we tend to avoid debug logs and go straight to trace